app/updatemanager.go

```diff
@@ -301,11 +319,66 @@ func (c *ControlMap) Set(cm *UpdateControlMap) {
 			return
 		}
 	}
+	// If the map exists in the expired pool, the map is removed from the
+	// expired pool
+	if UpdateManagerMapExpired.Contains(cm) {
+		UpdateManagerMapExpired = UpdateManagerMapExpired.Delete(cm)
+	}
+	// Make sure the Expired map has the same keys
+	UpdateManagerMapExpired.InsertKey(cm.ID)
 	c.controlMap[cm.ID] = append(c.controlMap[cm.ID], cm)
 }
 
 func (c *ControlMap) Get(ID string) []*UpdateControlMap {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
-	return c.controlMap[ID]
+	m := c.controlMap[ID]
+	// Prune the control map based on the UpdateCOntrolMapTimeout expiry
+	// FIXME - This should probably be handled some other way, like a
+	// separate goroutine (?)
+	for i, e := range m {
+		if e.Expired() {
+			// Moving to the expired pool...
+			UpdateManagerMapExpired.Set(e)
+			// Deleting from the active pool
+			m = append(m[:i], m[i+1:]...)
+		}
+	}
+	return m
+}
+
+func (c *ControlMap) Clear() {
+	c.controlMap = make(map[string][]*UpdateControlMap)
+}
+
+func (c *ControlMap) Delete(cm *UpdateControlMap) *ControlMap {
+	m, ok := c.controlMap[cm.ID]
+	if !ok {
+		return c
+	}
+	for i, e := range m {
+		if e == cm {
+			log.Debugf("Deleting %#v from the expired map", cm)
+			m = append(m[:i], m[i+1:]...)
+		}
+	}
+	return c
```

Comment:
With regards to our conversation on copy vs mutation.

This one is very strange to me: we are actually mutating (pointer receiver method) and then returning itself? Can you explain me the benefit?

Reply:
Nope, it's supposed to be a copy, fixing

Reply:
It was just a quick hack to get it to compile at the time :)